File: myHMM.R
myinitHMM <- function (States, Symbols, startProbs = NULL, transProbs = NULL, 
                       emissionProbs = NULL) 
{
  nStates <- length(States)
  nSymbols <- length(Symbols)
  SV <- rep(1/nStates, nStates)
  TM <- 0.5 * diag(nStates) + array(0.5/(nStates), c(nStates, nStates))
  EM <- array(1/(nSymbols), c(nStates, nSymbols))
  names(SV) <- States
  dimnames(TM) <- list(from = States, to = States)
  dimnames(EM) <- list(states = States, symbols = Symbols)
  if (!is.null(startProbs)) {
    SV[] <- startProbs[]
  }
  if (!is.null(transProbs)) {
    TM[, ] <- transProbs[, ]
  }
  if (!is.null(emissionProbs)) {
    EM[, ] <- emissionProbs[, ]
  }
  return(list(States = States, Symbols = Symbols, startProbs = SV, 
              transProbs = TM, emissionProbs = EM))
}


myforward <- function (hmm, observation) 
{
  nObservations <- length(observation)
  nStates <- length(hmm$States)
  hmm$transProbs[is.na(hmm$transProbs)] <- 0
  hmm$emissionProbs[is.na(hmm$emissionProbs)] <- 0
  f <- array(NA, c(nStates, nObservations))
  dimnames(f) <- list(states = hmm$States, index = 1:nObservations)
  for (state in hmm$States) {
    f[state, 1] <- log(hmm$startProbs[state] * 
                         hmm$emissionProbs[state, observation[1]])
  }
  for (k in 2:nObservations) {
    for (state in hmm$States) {
      logsum <- -Inf
      for (previousState in hmm$States) {
        temp <- f[previousState, k - 1] + 
          log(hmm$transProbs[previousState, state])
        logsum <- max(logsum, temp) + log(1 + exp(-abs(logsum - temp)))
      }
      f[state, k] <- log(hmm$emissionProbs[state, observation[k]]) + 
        logsum
    }
  }
  return(f)
}


mybackward <- function (hmm, observation) 
{
  nObservations <- length(observation)
  nStates <- length(hmm$States)
  hmm$transProbs[is.na(hmm$transProbs)] <- 0
  hmm$emissionProbs[is.na(hmm$emissionProbs)] <- 0
  b <- array(NA, c(nStates, nObservations))
  dimnames(b) <- list(states = hmm$States, index = 1:nObservations)
  for (state in hmm$States) {
    b[state, nObservations] <- log(1)
  }
  for (k in (nObservations - 1):1) {
    for (state in hmm$States) {
      logsum <- -Inf
      for (nextState in hmm$States) {
        temp <- b[nextState, k + 1] + 
          log(hmm$transProbs[state, nextState]) + 
          log(hmm$emissionProbs[nextState, observation[k + 1]])
        logsum <- max(logsum, temp) + log(1 + exp(-abs(logsum - temp)))
      }
      b[state, k] <- logsum
    }
  }
  return(b)
}


log_sum_exp <- function(x)
{
  max_x <- max(x)
  max_x + log(sum(exp(x - max_x)))
}


myBWR <- function (hmm, obs_list) 
{
  nOBSseq <- length(obs_list)
  nStates <- length(hmm$States)
  logstart_Y <- matrix(nrow = nOBSseq, ncol = nStates)
  dimnames(logstart_Y) <- list(index = 1:nOBSseq, states = hmm$States)
  TransitionMatrix <- hmm$transProbs
  TransitionMatrix[, ] <- 0
  logxi_list <- setNames(vector("list", nStates), hmm$States)
  logxi_list <- lapply(logxi_list, function(dummy) setNames(vector("list", nStates), hmm$States))
  EmissionMatrix <- hmm$emissionProbs
  EmissionMatrix[, ] <- 0
  loggamma_list <- setNames(vector("list", nStates), hmm$States)
  for (idx in 1:nOBSseq) {
    observation <- obs_list[[idx]]
    nObservations <- length(observation)
    f <- myforward(hmm, observation)
    b <- mybackward(hmm, observation)
    likelihood <- f[hmm$States[1], nObservations]
    for (state in hmm$States[-1]) {
      j <- f[state, nObservations]
      likelihood <- max(likelihood, j) + log(1 + exp(-abs(likelihood - j)))
    }
    for (state in hmm$States) {
      for (nextState in hmm$States) {
        for (i in 1:(nObservations - 1)) {
          logxi_list[[state]][[nextState]] <- c(logxi_list[[state]][[nextState]], 
            f[state, i] + log(hmm$transProbs[state, nextState]) + 
                  log(hmm$emissionProbs[nextState, observation[i + 1]]) + 
                  b[nextState, i + 1] - likelihood)
        }
      }
    }
    for (state in hmm$States) {
      loggamma_list[[state]] <- c(loggamma_list[[state]], 
                                  f[state, ] + b[state, ] - likelihood)
    }
    for (state in hmm$States) {
      logstart_Y[idx, state] <- f[state, 1] + b[state, 1] - likelihood
    }
  }
  for (state in hmm$States) {
    for (nextState in hmm$States) {
      TransitionMatrix[state, nextState] <- log_sum_exp(logxi_list[[state]][[nextState]]) - 
        log_sum_exp(unlist(logxi_list[[state]]))
    }
  }
  TransitionMatrix <- exp(TransitionMatrix)
  for (state in hmm$States) {
    for (s in hmm$Symbols) {
      EmissionMatrix[state, s] <- log_sum_exp(loggamma_list[[state]][unlist(obs_list) == s]) - 
        log_sum_exp(loggamma_list[[state]])
    }
  }
  EmissionMatrix <- exp(EmissionMatrix)
  return(list(TransitionMatrix = TransitionMatrix,
              EmissionMatrix = EmissionMatrix,
              startVec = exp(apply(logstart_Y, 2, log_sum_exp))))
}


myBW <- function (hmm, obs_list, maxIterations = 100, delta = 1e-9, 
                  pseudoCount = 0) 
{
  tempHmm <- hmm
  tempHmm$transProbs[is.na(hmm$transProbs)] <- 0
  tempHmm$emissionProbs[is.na(hmm$emissionProbs)] <- 0
  diff <- c()
  for (i in 1:maxIterations) {
    bw <- myBWR(tempHmm, obs_list)
    TM <- bw$TransitionMatrix
    EM <- bw$EmissionMatrix
    SV <- bw$startVec
    TM[!is.na(hmm$transProbs)] <- TM[!is.na(hmm$transProbs)] + pseudoCount
    EM[!is.na(hmm$emissionProbs)] <- EM[!is.na(hmm$emissionProbs)] + pseudoCount
    SV[!is.na(hmm$startProbs)] <- SV[!is.na(hmm$startProbs)] + pseudoCount
    TM <- (TM/rowSums(TM))
    EM <- (EM/rowSums(EM))
    SV <- (SV/sum(SV))
    d <- sqrt(sum((tempHmm$transProbs - TM)^2)) + 
      sqrt(sum((tempHmm$emissionProbs - EM)^2)) + 
      sqrt(sum((tempHmm$startProbs - SV)^2))
    diff <- c(diff, d)
    tempHmm$transProbs <- TM
    tempHmm$emissionProbs <- EM
    tempHmm$startProbs <- SV
    if (d < delta) {
      break
    }
  }
  tempHmm$transProbs[is.na(hmm$transProbs)] <- NA
  tempHmm$emissionProbs[is.na(hmm$emissionProbs)] <- NA
  tempHmm$startProbs[is.na(hmm$startProbs)] <- NA
  return(list(hmm = tempHmm, difference = diff))
}

File: myHMM.py
import numpy as np
import copy


def myinitHMM(states, symbols, start_probs=None, trans_probs=None, emission_probs=None):
    n_states = len(states)
    n_symbols = len(symbols)
    sv = np.full(n_states, 1.0 / n_states)
    tm = 0.5 * np.eye(n_states) + (0.5 / n_states) * np.ones((n_states, n_states))
    em = np.full((n_states, n_symbols), 1.0 / n_symbols)
    if start_probs is not None:
        sv = np.array(start_probs, dtype=float)
    if trans_probs is not None:
        tm = np.array(trans_probs, dtype=float)
    if emission_probs is not None:
        em = np.array(emission_probs, dtype=float)
    return {
        'states': states,
        'symbols': symbols,
        'start_probs': sv,
        'trans_probs': tm,
        'emission_probs': em
    }


def myforward(hmm, observation):
    n_obs = len(observation)
    n_states = len(hmm['states'])
    trans = np.nan_to_num(hmm['trans_probs'], nan=0.0)
    emis = np.nan_to_num(hmm['emission_probs'], nan=0.0)
    f = np.empty((n_states, n_obs))
    f[:] = np.nan
    for i in range(n_states):
        f[i, 0] = np.log(hmm['start_probs'][i] * emis[i, observation[0]])
    for k in range(1, n_obs):
        for i in range(n_states):
            logsum = -np.inf
            for j in range(n_states):
                temp = f[j, k-1] + np.log(trans[j, i])
                logsum = max(logsum, temp) + np.log1p(np.exp(-abs(logsum - temp)))
            f[i, k] = np.log(emis[i, observation[k]]) + logsum
    return f


def mybackward(hmm, observation):
    n_obs = len(observation)
    n_states = len(hmm['states'])
    trans = np.nan_to_num(hmm['trans_probs'], nan=0.0)
    emis = np.nan_to_num(hmm['emission_probs'], nan=0.0)
    b = np.empty((n_states, n_obs))
    for i in range(n_states):
        b[i, n_obs - 1] = 0.0
    for k in range(n_obs - 2, -1, -1):
        for i in range(n_states):
            logsum = -np.inf
            for j in range(n_states):
                temp = (
                    b[j, k + 1]
                    + np.log(trans[i, j])
                    + np.log(emis[j, observation[k + 1]])
                )
                logsum = max(logsum, temp) + np.log1p(np.exp(-abs(logsum - temp)))
            b[i, k] = logsum
    return b



def log_sum_exp(x):
    max_x = np.max(x)
    return max_x + np.log(np.sum(np.exp(x - max_x)))


def myBWR(hmm, obs_list):
    n_seq = len(obs_list)
    n_states = len(hmm['states'])
    n_symbols = len(hmm['symbols'])
    logstart_Y = np.empty((n_seq, n_states))
    TransitionMatrix = np.zeros((n_states, n_states))
    logxi_list = [[[] for _ in range(n_states)] for _ in range(n_states)]
    loggamma_list = [[] for _ in range(n_states)]
    EmissionMatrix = np.zeros((n_states, n_symbols))
    for idx, observation in enumerate(obs_list):
        n_obs = len(observation)
        f = myforward(hmm, observation)
        b = mybackward(hmm, observation)
        likelihood = log_sum_exp(f[:, n_obs-1])
        for i in range(n_states):
            for j in range(n_states):
                for t in range(n_obs-1):
                    logxi_list[i][j].append(
                        f[i, t]
                        + np.log(hmm['trans_probs'][i, j])
                        + np.log(hmm['emission_probs'][j, observation[t+1]])
                        + b[j, t+1]
                        - likelihood
                    )
        for i in range(n_states):
            gamma_vals = f[i, :] + b[i, :] - likelihood
            loggamma_list[i].extend(gamma_vals.tolist())
        for i in range(n_states):
            logstart_Y[idx, i] = f[i, 0] + b[i, 0] - likelihood
    for i in range(n_states):
        denom = log_sum_exp(np.array([xi for row in logxi_list[i] for xi in row]))
        for j in range(n_states):
            num = log_sum_exp(np.array(logxi_list[i][j]))
            TransitionMatrix[i, j] = np.exp(num - denom)
    obs_flat = [sym for seq in obs_list for sym in seq]
    for i in range(n_states):
        gamma_flat = loggamma_list[i]
        denom = log_sum_exp(np.array(gamma_flat))
        for s_idx in range(n_symbols):
            positions = [k for k, sym in enumerate(obs_flat) if sym == s_idx]
            if positions:
                vals = [gamma_flat[k] for k in positions]
                EmissionMatrix[i, s_idx] = np.exp(log_sum_exp(np.array(vals)) - denom)
    startVec = np.exp([log_sum_exp(logstart_Y[:, i]) for i in range(n_states)])
    return {'TransitionMatrix': TransitionMatrix,
            'EmissionMatrix': EmissionMatrix,
            'startVec': startVec}


def myBW(hmm, obs_list, maxIterations=100, delta=1e-9, pseudoCount=0):
    tempHmm = copy.deepcopy(hmm)
    mask_trans = ~np.isnan(hmm['trans_probs'])
    mask_emis = ~np.isnan(hmm['emission_probs'])
    mask_sv = ~np.isnan(hmm['start_probs'])
    diff = []
    for _ in range(maxIterations):
        bw = myBWR(tempHmm, obs_list)
        TM = bw['TransitionMatrix']
        EM = bw['EmissionMatrix']
        SV = bw['startVec']
        TM[mask_trans] += pseudoCount
        EM[mask_emis] += pseudoCount
        SV[mask_sv] += pseudoCount
        TM = TM / TM.sum(axis=1)[:, None]
        EM = EM / EM.sum(axis=1)[:, None]
        SV = SV / SV.sum()
        d = (np.sqrt(np.sum((tempHmm['trans_probs'] - TM)**2)) +
             np.sqrt(np.sum((tempHmm['emission_probs'] - EM)**2)) +
             np.sqrt(np.sum((tempHmm['start_probs'] - SV)**2)))
        diff.append(d)
        tempHmm['trans_probs'] = TM.copy()
        tempHmm['emission_probs'] = EM.copy()
        tempHmm['start_probs'] = SV.copy()
        if d < delta:
            break
    tempHmm['trans_probs'][~mask_trans] = np.nan
    tempHmm['emission_probs'][~mask_emis] = np.nan
    tempHmm['start_probs'][~mask_sv] = np.nan
    return {'hmm': tempHmm, 'difference': diff}
